src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -198,20 +195,30 @@ private void OnWriteCompleted(Queue<ArraySegment<byte>> writtenBuffers, int stat
                     // completed writes that we haven't triggered callbacks for yet.
                     _numBytesPreCompleted -= writeBuffer.Count;
                 }
-
-
+                
                 // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
                 // This allows large writes to complete once they've actually finished.
                 var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
-                while (_callbacksPending.Count > 0 &&
-                       _callbacksPending.Peek().BytesToWrite <= bytesLeftToBuffer)
+                while (_tasksPending.Count > 0 &&
+                       (int)(_tasksPending.Peek().Task.AsyncState) <= bytesLeftToBuffer)
                 {
-                    var callbackContext = _callbacksPending.Dequeue();
+                    var tcs = _tasksPending.Dequeue();
 
-                    _numBytesPreCompleted += callbackContext.BytesToWrite;
+                    _numBytesPreCompleted += (int)(tcs.Task.AsyncState);
 
-                    // callback(error, state, calledInline)
-                    callbackContext.Callback(_lastWriteError, callbackContext.State, false);
+                    if (error == null)
+                    {
+                        ThreadPool.QueueUserWorkItem(
+                            (o) => ((TaskCompletionSource<object>)o).SetResult(null), 
+                            tcs);
+                    }
+                    else
+                    {
+                        // error is closure captured 
+                        ThreadPool.QueueUserWorkItem(
+                            (o) => ((TaskCompletionSource<object>)o).SetException(error), 
+                            tcs);
+                    }
                 }
 
                 // Now that the while loop has completed the following invariants should hold true:
```

Comment:
It's hidden in this PR, but I don't think `Debug.Assert(_numBytesPreCompleted <= _maxBytesPreCompleted);` is still valid since we can now exceed that given non-immediate writes.

I would just remove that assertion.